log/gokit.go

```diff
@@ -6,101 +6,93 @@ package log
 
 import (
 	"fmt"
+	"io"
 	"os"
 
 	"github.com/go-kit/log"
 	"github.com/go-kit/log/level"
 )
 
-// NewGoKitFormat creates a new Interface backed by a GoKit logger
-// format can be "json" or defaults to logfmt
-func NewGoKitFormat(l Level, f Format) Interface {
-	var logger log.Logger
-	if f.s == "json" {
-		logger = log.NewJSONLogger(log.NewSyncWriter(os.Stderr))
-	} else {
-		logger = log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr))
-	}
-	return addStandardFields(logger, l)
-}
-
-// stand-alone for test purposes
-func addStandardFields(logger log.Logger, l Level) Interface {
-	logger = log.With(logger, "ts", log.DefaultTimestampUTC, "caller", log.Caller(5))
-	logger = level.NewFilter(logger, l.Gokit)
-	return gokit{logger}
-}
-
-// NewGoKit creates a new Interface backed by a GoKit logger
-func NewGoKit(l Level) Interface {
-	return NewGoKitFormat(l, Format{s: "logfmt"})
-}
-
-// GoKit wraps an existing gokit Logger.
-func GoKit(logger log.Logger) Interface {
-	return gokit{logger}
-}
+const (
+	LogfmtFormat = "logfmt"
+)
 
-type gokit struct {
-	log.Logger
-}
+type Fields []interface{}
 
-// Helper to defer sprintf until it is needed.
-type sprintf struct {
-	format string
-	args   []interface{}
+func NewFields(fields ...interface{}) Fields {
+	return fields
 }
 
-func (s *sprintf) String() string {
-	return fmt.Sprintf(s.format, s.args...)
+func (f Fields) empty() bool {
+	return len(f) == 0
 }
 
-// Helper to defer sprint until it is needed.
-// Note we don't use Sprintln because the output is passed to go-kit as one value among many on a line
-type sprint struct {
-	args []interface{}
+// NewGoKit creates a new GoKit logger with the given level, format and writer,
+// enriched with the standard "ts" and "caller" fields.
+// If the given writer is nil, os.Stderr is used.
+// If the given format is nil, logfmt is used.
+func NewGoKit(lvl Level, format string, writer io.Writer) log.Logger {
+	logger := newGoKit(format, writer, nil)
+	logger = addStandardFields(logger)
+	return level.NewFilter(logger, lvl.Option)
 }
 
-func (s *sprint) String() string {
-	return fmt.Sprint(s.args...)
+// NewGoKitLogFmt creates a new GoKit logger with the given level and writer,
+// enriched with the standard "ts" and "caller" fields.
+// The "logfmt" format is used.
+// If the given writer is nil, os.Stderr is used.
+func NewGoKitLogFmt(l Level, writer io.Writer) log.Logger {
+	return NewGoKit(l, "logfmt", writer)
```

Comment:
[nit] I doubt we need this function at all. I think the caller would just call `NewGoKit()` passing `LogfmtFormat`. If you decide to keep it, then see the following suggestion

```suggestion
	return NewGoKit(l, LogfmtFormat, writer)
```

Reply:
Removed